File: sdks/typescript/src/WideEvent.ts
import { performance } from 'node:perf_hooks';
import { randomUUID } from 'node:crypto';

import type { EventStatus, LogLevel, WideEventPayload, Sink } from './types.js';
import { cloneStructured, deepFreeze, normalizeEnvironment } from './internal/snapshot.js';
import { SDK_VERSION } from './version.js';

const DEFAULT_ERROR = { type: null, message: null, stack: null };

function normalizeError(err: Error | unknown): WideEventPayload['error'] {
  if (err instanceof Error) {
    return {
      type: err.constructor?.name ?? 'Error',
      message: err.message || null,
      stack: err.stack ?? null,
    };
  }

  if (typeof err === 'string') {
    return { type: 'Error', message: err, stack: null };
  }

  if (err === null || err === undefined) {
    return { ...DEFAULT_ERROR };
  }

  return {
    type: typeof err,
    message: (() => {
      try {
        return JSON.stringify(err);
      } catch {
        return String(err);
      }
    })(),
    stack: null,
  };
}

export class WideEvent {
  private readonly event_id = randomUUID();
  private readonly timestamp_start = new Date();
  private readonly start_monotonic_ms = performance.now();
  private readonly sinks: readonly Sink[];
  private readonly service: string;
  private readonly environment: ReturnType<typeof normalizeEnvironment>;
  private context: Record<string, unknown> = {};
  private error: WideEventPayload['error'] = { ...DEFAULT_ERROR };
  private level: LogLevel = 'info';
  private status: EventStatus = 'ok';
  private finalized = false;
  private emitPromise: Promise<void> | null = null;

  constructor(
    private readonly operation: string,
    service: string,
    environment: string,
    sinks: Sink[],
  ) {
    this.service = service;
    this.environment = normalizeEnvironment(environment);
    this.sinks = [...sinks];
  }

  private assertMutable() {
    if (this.finalized) {
      throw new Error('WideEvent has already been finalized');
    }
  }

  /** Merge arbitrary key-value context into the event. Chainable. */
  appendContext(fields: Record<string, unknown>): this {
    this.assertMutable();
    Object.assign(this.context, cloneStructured(fields));
    return this;
  }

  /** Set structured error. Chainable. */
  setError(err: Error | unknown): this {
    this.assertMutable();
    this.level = 'error';
    this.status = 'error';
    this.error = normalizeError(err);
    return this;
  }

  private buildPayload(overrides?: Partial<Pick<WideEventPayload, 'level' | 'status'>>): WideEventPayload {
    const timestamp_end = new Date();
    const payload: WideEventPayload = {
      event_id: this.event_id,
      service: this.service,
      environment: this.environment,
      timestamp_start: this.timestamp_start.toISOString(),
      timestamp_end: timestamp_end.toISOString(),
      duration_ms: Number((performance.now() - this.start_monotonic_ms).toFixed(3)),
      level: overrides?.level ?? this.level,
      operation: this.operation,
      status: overrides?.status ?? this.status,
      context: cloneStructured(this.context),
      error: cloneStructured(this.error),
      meta: { sdk_version: SDK_VERSION, sdk_language: 'typescript' },
    };

    return deepFreeze(payload);
  }

  /** Finalizes the event and dispatches to all sinks. */
  async emit(overrides?: Partial<Pick<WideEventPayload, 'level' | 'status'>>): Promise<void> {
    if (this.emitPromise) {
      return this.emitPromise;
    }

    const payload = this.buildPayload(overrides);
    this.finalized = true;

    this.emitPromise = Promise.allSettled(
      this.sinks.map((sink) => Promise.resolve().then(() => sink.export(payload))),
    ).then(() => undefined);

    return this.emitPromise;
  }
}

File: sdks/python/src/openakta_logger/wide_event.py
from __future__ import annotations

import asyncio
import copy
import inspect
import json
import time
import traceback
import uuid
from datetime import datetime, timezone
from typing import Any

from .sinks.base import Sink

SDK_VERSION = "0.1.0"
VALID_ENVIRONMENTS = {"production", "staging", "development"}


def _normalize_environment(value: str | None) -> str:
    if value in VALID_ENVIRONMENTS:
        return value
    return "production"


def _normalize_error(exc: BaseException | Any) -> dict[str, str | None]:
    if isinstance(exc, BaseException):
        return {
            "type": type(exc).__name__,
            "message": str(exc) or None,
            "stack": "".join(traceback.format_exception(type(exc), exc, exc.__traceback__)),
        }

    if exc is None:
        return {"type": None, "message": None, "stack": None}

    if isinstance(exc, str):
        return {"type": "Error", "message": exc, "stack": None}

    try:
        message = json.dumps(exc, default=str)
    except Exception:
        message = str(exc)

    return {"type": type(exc).__name__, "message": message, "stack": None}


class WideEvent:
    def __init__(
        self,
        operation: str,
        service: str,
        environment: str,
        sinks: list[Sink],
    ) -> None:
        self._event_id = str(uuid.uuid4())
        self._timestamp_start = datetime.now(timezone.utc)
        self._start_monotonic = time.perf_counter()
        self._operation = operation
        self._service = service
        self._environment = _normalize_environment(environment)
        self._sinks = list(sinks)
        self._context: dict[str, Any] = {}
        self._error: dict[str, str | None] = {"type": None, "message": None, "stack": None}
        self._level = "info"
        self._status = "ok"
        self._finalized = False
        self._emit_task: asyncio.Task[None] | None = None

    def _assert_mutable(self) -> None:
        if self._finalized:
            raise RuntimeError("WideEvent has already been finalized")

    def append_context(self, **fields: Any) -> "WideEvent":
        """Merge arbitrary fields into the event context."""
        self._assert_mutable()
        self._context.update(copy.deepcopy(fields))
        return self

    def set_error(self, exc: BaseException | Any) -> "WideEvent":
        self._assert_mutable()
        self._level = "error"
        self._status = "error"
        self._error = _normalize_error(exc)
        return self

    def _build_payload(
        self,
        level: str | None = None,
        status: str | None = None,
    ) -> dict[str, Any]:
        timestamp_end = datetime.now(timezone.utc)
        payload = {
            "event_id": self._event_id,
            "service": self._service,
            "environment": self._environment,
            "timestamp_start": self._timestamp_start.isoformat(),
            "timestamp_end": timestamp_end.isoformat(),
            "duration_ms": round((time.perf_counter() - self._start_monotonic) * 1000, 3),
            "level": level or self._level,
            "operation": self._operation,
            "status": status or self._status,
            "context": copy.deepcopy(self._context),
            "error": copy.deepcopy(self._error),
            "meta": {"sdk_version": SDK_VERSION, "sdk_language": "python"},
        }
        return payload

    async def _safe_export(self, sink: Sink, payload: dict[str, Any]) -> None:
        try:
            await sink.export(payload)
        except Exception:
            return

    async def _dispatch(self, payload: dict[str, Any]) -> None:
        await asyncio.gather(*(self._safe_export(sink, payload) for sink in self._sinks))

    async def emit(self, level: str | None = None, status: str | None = None) -> None:
        if self._emit_task is not None:
            await self._emit_task
            return

        payload = self._build_payload(level=level, status=status)
        self._finalized = True
        self._emit_task = asyncio.create_task(self._dispatch(payload))
        await self._emit_task
